### PR TITLE
Add constant FilesystemIterator::SKIP_DOTS when flags parameter is used

### PR DESCRIPTION
**Pull request by jawira**
I created a new rector rule to update `\FilesystemIterator` , this rule is supposed to be used when migrating code from `php 8.1` to `8.2`.

Before PHP 8.2, `\FilesystemIterator::SKIP_DOTS` was always set and couldn't be removed, this has been fixed in PHP 8.2 and now you have to add this constant manually if you want to maintain the same behavior. This is also explained in documentation <https://www.php.net/manual/en/filesystemiterator.construct.php>.

You can see an example here <https://3v4l.org/HrmZ2>.

Note that this rule is important only if you use the second parameter (`$params`) from `FilesystemIterator`, the reason is that `\FilesystemIterator::SKIP_DOTS` is part of `$params `default value.

```php
// Nothing to do since \FilesystemIterator::SKIP_DOTS is part of default value.  
new \FilesystemIterator(__DIR__);

// Must add \FilesystemIterator::SKIP_DOTS to maintain same behavior before PHP 8.2
new \FilesystemIterator(__DIR__, \FilesystemIterator::KEY_AS_FILENAME);
``` 

I don't have experience creating Rector rules, so I'm open to any comment/suggestion :)

### CODE REVIEW

File: rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            ]
+        );
+    }
+
+    public function getNodeTypes(): array
+    {
+        return [New_::class];
+    }
+
+    public function refactor(Node $node): ?New_
```
Comment (samsonasik) [inline rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php:55]: you can use `@param New_` like other rule:

```suggestion
    /**
     * @param New_ $node
     */
    public function refactor(Node $node): ?New_
```

File: rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    {
+        assert($node instanceof New_);
```
Comment (samsonasik) [inline rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php]: this assert is not needed:

```suggestion
```

File: rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        if (! array_key_exists(1, $node->args)) {
```
Comment (samsonasik) [inline rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php]: is first class callable check is needed:

```suggestion
        if ($node->isFirstClassCallable()) {
            return null;
        }
        
        if (! array_key_exists(1, $node->args)) {
```

File: build/target-repository/docs/rector_rules_overview.md [hunk 4]
```diff
@@ -6506,6 +6506,19 @@ Refactor Spatie enum class to native Enum
...
+Prior PHP 8.2 FilesystemIterator::SKIP_DOTS was always set and could not be removed, therefore FilesystemIterator::SKIP_DOTS is added in order to keep this behaviour.
+
+- class: [`Rector\Php82\Rector\New_\FilesystemIteratorSkipDotsRector`](../rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php)
+
+```diff
+-new \FilesystemIterator(__DIR__, \FilesystemIterator::KEY_AS_FILENAME);
++new \FilesystemIterator(__DIR__, \FilesystemIterator::KEY_AS_FILENAME | \FilesystemIterator::SKIP_DOTS);
+```
+
+<br>
```
Comment (samsonasik) [inline build/target-repository/docs/rector_rules_overview.md:6584]: documentation needs rebase, you can run:

```
git pull --rebase https://github.com/rectorphp/rector-src.git main
composer docs
git add .
git rebase --continue
git push --force origin filesystem_iterator_skip_dots 
```

File: rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+    public function provideMinPhpVersion(): int
+    {
+        return PhpVersionFeature::FILESYSTEM_ITERATOR_SKIP_DOTS;
+    }
+
+    /**
+     * Is the constant {@see \FilesystemIterator::SKIP_DOTS} present?
+     */
+    private function isSkipDotsPresent(Expr $node): bool
```
Comment (samsonasik) [inline rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php:80]: I prefer to verify the expr is ConstFetch or `BitwiseOr`, so verify the target ConstFetch exists in the `BitwiseOr`

Comment (jawira) [reply rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php:80]: @samsonasik Hello, I have made the requested change using ClassConstFetch, but I also made a mistake while rebasing and now there's +100 files in this pull request :(
I'm thinking about closing this PR and create a new one, what's your opinion?

Comment (samsonasik) [reply rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php:80]: I wrote a blog post about how to rebase https://samsonasik.wordpress.com/2015/09/16/practical-git-4-rebasing-conflicted-task-branch-against-primary-branch/

Comment (jawira) [reply rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php:80]: Hello, I applied requested changes :+1:

File: build/target-repository/docs/rector_rules_overview.md (hunk 4 above)
Comment (jawira) [reply build/target-repository/docs/rector_rules_overview.md:6584]: Hello, I rebased and updted doc

File: rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php (hunk 5 above)
Comment (samsonasik) [reply rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php:80]: could you rebase latest main branch?

Comment (jawira) [reply rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php:80]: Hello, I just rebased the branch

File: rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    {
+        $element = $this->betterNodeFinder->findFirst($node, static function (Node $node): bool {
+            if (! $node instanceof ClassConstFetch) {
+                return false;
+            }
+            return strval($node->name) === 'SKIP_DOTS';
+        });
```
Comment (samsonasik) [inline rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php]: You can use `BitwiseOr` node which has Collection of ClassConstFetch and only ClassConstFetch

Comment (jawira) [reply rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php]: @samsonasik Hello, I don't understand what the problem is, can you elaborate please ?

Comment (samsonasik) [reply rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php]: You can check that the args[1] is :+1: 

- BitwiseOr
- ClassConstFetch

if it is `BitwiseOr`, loop its `left` node until its only a `ClassConstFetch`, if you found a non-ClassConstFetch under `BitwiseOr`, something like:

```
        $left = $bitwiseOr->left;
        while ($left instanceof BitwiseOr) {
            if (! $left instanceof BitwiseOr && ! $left instanceof ClassConstFetch) {
                     // skip
                     return true;
            }

            // verify ClassConstFetch has $name is Identifier
           if (! $classConstFetch->name instanceof Identifier) {
              // skip
              return true;
          } 

          if (! $classConstFetch->class instanceof Name) {
               // skip 
                return true;
           }

           if (! $this->isName($classConstFetch->name, 'SKIP_DOTS')) {
                 // re-assign to continue loop left Node parts
                 $left = $left->left;
                 continue;
           }

           if (! $this->isName($classConstFetch->class, 'FilesystemIterator')) {
                 // skip, can be any class and any other class const fetch
                 true;
           }

            return false;
        }
```

above is pseudo code, so possibly cause infinite loop, use with verify on your local devs :)

File: rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+        return false;
+    }
+
+    /**
+     * Tells if $expr is equal to {@see \FilesystemIterator::SKIP_DOTS}.
+     */
+    protected function isSkipDots(Expr $expr): bool
+    {
+        return ($expr instanceof ClassConstFetch) && (strval($expr->name) === 'SKIP_DOTS');
```
Comment (samsonasik) [inline rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php]: You may can use `ClassConstFetchNameResolver::resolve($expr)` and verify if it equal to : `FilesystemIterator::SKIP_DOTS`

File: rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/constant_present_with_fqn.php.inc [hunk 8]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+namespace Rector\Tests\Php82\Rector\New_\FilesystemIteratorSkipDots\Fixture;
+
+final class ConstantPresentWithFqn
```
Comment (samsonasik) [inline rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/constant_present_with_fqn.php.inc]:
```suggestion
final class SkipConstantPresentWithFqn
```

File: rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/constant_present_in_flags.php.inc [hunk 9]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+namespace Rector\Tests\Php82\Rector\New_\FilesystemIteratorSkipDots\Fixture;
+
+use FilesystemIterator;
+
+final class ConstantPresentInFlags
```
Comment (samsonasik) [inline rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/constant_present_in_flags.php.inc]:
```suggestion
final class SkipConstantPresentInFlags
```

File: rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/constant_present.php.inc [hunk 10]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+namespace Rector\Tests\Php82\Rector\New_\FilesystemIteratorSkipDots\Fixture;
+
+use FilesystemIterator;
+
+final class ConstantPresent
```
Comment (samsonasik) [inline rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/constant_present.php.inc]:
```suggestion
final class SkipConstantPresent
```

File: rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/constant_present.php.inc [hunk 11]
```diff
@@ -0,0 +1,15 @@
+<?php
```
Comment (samsonasik) [inline rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/constant_present.php.inc]: please rename fixture file name to : `skip_constant_present.php.inc`

File: rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/constant_present_in_flags.php.inc [hunk 12]
```diff
@@ -0,0 +1,15 @@
+<?php
```
Comment (samsonasik) [inline rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/constant_present_in_flags.php.inc]: please rename fixture file name to : `skip_constant_present_in_flags.php.inc`

File: rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/constant_present_with_fqn.php.inc [hunk 13]
```diff
@@ -0,0 +1,13 @@
+<?php
```
Comment (samsonasik) [inline rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/constant_present_with_fqn.php.inc]: please rename fixture file name to : `skip_constant_present_with_fqn.php.inc`

File: rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/maintain_default_value.php.inc [hunk 14]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+namespace Rector\Tests\Php82\Rector\New_\FilesystemIteratorSkipDots\Fixture;
+
+use FilesystemIterator;
+
+final class MaintainDefaultValue
```
Comment (samsonasik) [inline rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/maintain_default_value.php.inc]:
```suggestion
final class SkipWithDefaultValue
```

File: rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/maintain_default_value.php.inc [hunk 15]
```diff
@@ -0,0 +1,15 @@
+<?php
```
Comment (samsonasik) [inline rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/maintain_default_value.php.inc]: please rename fixture file to `skip_with_default_value.php.inc`

File: rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php (hunk 7 above)
Comment (jawira) [reply rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php]: I'm currently using ClassConstFetchNameResolver

File: rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php (hunk 6 above)
Comment (jawira) [reply rules/Php82/Rector/New_/FilesystemIteratorSkipDotsRector.php]: I refactored code and I'm using while loop now.

File: rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/constant_present_with_fqn.php.inc (hunk 8 above)
Comment (jawira) [reply rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/constant_present_with_fqn.php.inc]: I renamed the class

File: rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/constant_present_in_flags.php.inc (hunk 9 above)
Comment (jawira) [reply rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/constant_present_in_flags.php.inc]: I renamed the class

File: rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/constant_present.php.inc (hunk 10 above)
Comment (jawira) [reply rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/constant_present.php.inc]: I renamed the class

File: rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/constant_present.php.inc (hunk 11 above)
Comment (jawira) [reply rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/constant_present.php.inc]: I renamed the file

File: rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/constant_present_in_flags.php.inc (hunk 12 above)
Comment (jawira) [reply rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/constant_present_in_flags.php.inc]: I renamed the file

File: rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/constant_present_with_fqn.php.inc (hunk 13 above)
Comment (jawira) [reply rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/constant_present_with_fqn.php.inc]: I renamed the file

File: rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/maintain_default_value.php.inc (hunk 14 above)
Comment (jawira) [reply rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/maintain_default_value.php.inc]: I renamed the class

File: rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/maintain_default_value.php.inc (hunk 15 above)
Comment (jawira) [reply rules-tests/Php82/Rector/New_/FilesystemIteratorSkipDots/Fixture/maintain_default_value.php.inc]: I renamed the file

**Review by samsonasik — APPROVED** [review]
Looks good 👍

**Comment by TomasVotruba** [comment]
Thank you :clap:

**Comment by samsonasik** [comment]
Downgrade build failure seems unrelated:

```diff

------------------------------------------------------------
Parse error: rector-prefixed-downgraded/vendor/nette/utils/src/Utils/Arrays.php:171
    169|      */
    170|     public static function grep(array $array, #[\JetBrains\PhpStorm\Language('RegExp')] string $pattern, int $flags = 0) : array
  > 171|     {
    172|         return Strings::pcre('preg_grep', [$pattern, $array, $flags]);
    173|     }
Unexpected '{', expecting variable (T_VARIABLE) in rector-prefixed-downgraded/vendor/nette/utils/src/Utils/Arrays.php on line 171
------------------------------------------------------------
Parse error: rector-prefixed-downgraded/vendor/nette/utils/src/Utils/Strings.php:399
    397|      */
    398|     public static function split(string $subject, #[\JetBrains\PhpStorm\Language('RegExp')] string $pattern, int $flags = 0) : array
  > 399|     {
    400|         return self::pcre('preg_split', [$pattern, $subject, -1, $flags | \PREG_SPLIT_DELIM_CAPTURE]);
    401|     }
Unexpected '{', expecting variable (T_VARIABLE) in rector-prefixed-downgraded/vendor/nette/utils/src/Utils/Strings.php on line 399
```

seems due to new nette/utils release https://github.com/nette/utils/releases/tag/v3.2.9

**Comment by samsonasik** [comment]
The error seems on downgrading to php 7.4 on Attribute in parameter, ref https://3v4l.org/F7j9M#v7.4.33, which may solvable to re-print the parameter to make new line instead of single line https://3v4l.org/o82Pl#v7.4.2

**Comment by samsonasik** [comment]
Actually, there is already `DowngradeAttributeToAnnotationRector`, since it prefixed vendor that we will not use in directly outside rector, I think we can add more register to `config/set/downgrade-php80.php` in rector-downgrade-php

**Comment by samsonasik** [comment]
@TomasVotruba I created PR https://github.com/rectorphp/rector-downgrade-php/pull/36 for it.

**Comment by samsonasik** [comment]
@TomasVotruba I created another PR alternative specific rule to reprint the node when attribute line is equal with node line at PR https://github.com/rectorphp/rector-downgrade-php/pull/37

**Pull request merged**